src/cdk/overlay/keyboard/overlay-keyboard-dispatcher.ts

```diff
@@ -0,0 +1,96 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Injectable, Optional, SkipSelf, OnDestroy} from '@angular/core';
+import {OverlayRef} from '../overlay-ref';
+import {Subject} from 'rxjs/Subject';
+import {RxChain, filter, takeUntil} from '@angular/cdk/rxjs';
+import {fromEvent} from 'rxjs/observable/fromEvent';
+
+/**
+ * Service for dispatching keyboard events that land on the body to appropriate overlay ref,
+ * if any. It maintains a list of attached overlays to determine best suited overlay based
+ * on event target and order of overlay opens.
+ */
+@Injectable()
+export class OverlayKeyboardDispatcher implements OnDestroy {
+
+  /** Currently attached overlays in the order they were attached. */
+  _attachedOverlays: OverlayRef[] = [];
+
+  /** Emits when the service is destroyed. */
+  private _onDestroy = new Subject<void>();
+
+  constructor() {
+    this._dispatchKeydownEvents();
+  }
+
+  ngOnDestroy() {
+    this._onDestroy.next();
+    this._onDestroy.complete();
+  }
+
+  /** Add a new overlay to the list of attached overlay refs. */
+  add(overlayRef: OverlayRef): void {
+    this._attachedOverlays.push(overlayRef);
+  }
+
+  /** Remove an overlay from the list of attached overlay refs. */
+  remove(overlayRef: OverlayRef): void {
+    const index = this._attachedOverlays.indexOf(overlayRef);
+    if (index > -1) {
+      this._attachedOverlays.splice(index, 1);
+    }
+  }
+
+  /**
+   * Subscribe to keydown events that land on the body and dispatch those
+   * events to the appropriate overlay.
+   */
+  private _dispatchKeydownEvents(): void {
+    const bodyKeydownEvents = fromEvent<KeyboardEvent>(document.body, 'keydown');
+
+    RxChain.from(bodyKeydownEvents)
+      .call(filter, () => !!this._attachedOverlays.length)
+      .call(takeUntil, this._onDestroy)
+      .subscribe(event => {
+        // Dispatch keydown event to correct overlay reference
+        this._selectOverlayFromEvent(event)._dispatchedKeydownEvents.next(event);
+      });
+  }
+
+  /** Select the appropriate overlay from a keydown event. */
+  private _selectOverlayFromEvent(event: KeyboardEvent): OverlayRef {
+    // Check if any overlays contain the event
+    const targetedOverlay = this._attachedOverlays.find(overlay => {
+      return overlay.overlayElement.contains(event.target as HTMLElement) ||
+          overlay.overlayElement === event.target;
+    });
+
+    // Use that overlay if it exists, otherwise choose the most recently attached one
+    return targetedOverlay ?
```

Comment:
`return targetedOverlay || this._attachedOverlays[this._attachedOverlays.length - 1]`.